Nox stuff

Endercheif opened the pull request:
<!-- Please complete the missing ... parts. -->

### Changes

It has come to my attention that many of the measure we have in place are not being run or used as much. Nox will fix that. 
Before doing a pr, on can run `nox` on their local machine to run tests and lints. As a workflow, they will also run.

 ### Check off the following

-   [x] I have tested my changes with the current requirements
-   [x] My Code follows the pep8 code style.

Arthurdw commented on pyproject.toml:
```diff
@@ -74,5 +75,5 @@ warn_unused_configs = true
 
 [tool.black]
-line-length = 80
+line-length =  88
```
Any reason for this change?

Endercheif replied:
In .flake8 and .editorconfig it is set to 88

Sigmanificient commented:
What is tox doing?

Endercheif commented:
> What is tox doing?

tox (thing I removed) did nothing because it was never integrated into the code base, it was an extra dependancy. nox (what I added) runs black, isort, flake8, pytest, mypy all in one command. Prior to this, it was clear many had not ran these individually as they were supposed to. nox makes it easy and will also auto run as a GitHub workflow

Sigmanificient commented:
> > What is tox doing?
> 
> tox (thing I removed) did nothing because it was never integrated into the code base, it was an extra dependancy. nox (what I added) runs black, isort, flake8, pytest, mypy all in one command. Prior to this, it was clear many had not ran these individually as they were supposed to. nox makes it easy and will also auto run as a GitHub workflow

We maybe dont want to run mypy cause there is a lot of type errors currently

Sigmanificient commented:
nox still failling xD

Lunarmagpie commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,23 @@
+      - name: Python Setup
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
```
Nox should be run on all supported versions of python. You should use the matrix thing. Heres an example from one of my projects https://github.com/magpie-dev/hikari-crescent/blob/main/.github/workflows/pytest.yml

Lunarmagpie commented on noxfile.py:
```diff
@@ -0,0 +1,27 @@
+
+
+@nox.session
+def mypy(session: nox.Session):
+    session.install("mypy")
+    session.run("mypy", "pincer")
```
Mypy shouldn't be run. Theres far too many type issues so its always gonna fail.

Sigmanificient replied:
Yes that what i was saying too, we should run it later

Sigmanificient commented on noxfile.py:
```diff
@@ -0,0 +1,27 @@
+    session.run("flake8", "pincer")
+
+
+@nox.session
+def mypy(session: nox.Session):
+    session.install("mypy")
+    session.run("mypy", "pincer")
```
```suggestion
"""
Mypy will be be ran within a workflow until pincer provide complete typing
that will not fail for every push.

@nox.session
def mypy(session: nox.Session):
   session.install("mypy")
   session.run("mypy", "pincer")
"""
```

Sigmanificient replied:
@Endercheif

Sigmanificient commented on pyproject.toml:
```diff
@@ -25,7 +25,7 @@ classifiers = [
 
 [tool.poetry.dependencies]
-python = ">=3.8"
+python = ">=3.8,<3.11"
```
Breaking changes on python 3.11?

Arthurdw replied:
I don't think there are breaking changes for us.

Sigmanificient replied:
Why changing this then?

Endercheif replied:
@Sigmanificient @Arthurdw Poetry is very picky about supported version of dependencies. One such dependency (Nox I believe) does not support anything greater than 3.11 so the install fails. Unless this is added.

Sigmanificient commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,23 @@
+
+jobs:
+  ci:
+    runs-on: ${{ matrix.os }}-latest
+    strategy:
+      matrix:
+        os: [ ubuntu, windows, macos ]
+
+    name: Nox on ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v2
+      - name: Python Setup
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
```
@Endercheif

Sigmanificient commented:
Can we just format the coe with black one and for all?

Endercheif commented:
The only reason the Nox is failing, is because the code is not formatted correctly. 
We have two options. 
- Merge this now and in a different PR resolve the formatting issue or 
- In this PR, format it.

I would say the second one just to keep the PRs organized.

Arthurdw reviewed (approved):
Seems okay to me.
